docs/design/assembly-loading.md

```diff
@@ -0,0 +1,83 @@
+
+### Additional processing done lazily
+
+Additional processing will be done lazily (not triggered by `Resolve`), running on a single assembly at a time.
+
+- Type info base/override methods will be built per assembly as requested
+
+  The overrides tracked for a method may be incomplete, and will be updated as new assemblies are loaded. This is compatible with the current design of `MarkStep` which re-processes virtual methods, but this could be optimized in the future.
+
+- XML from an assembly will be processed only if the assembly is marked
+
+  Embedded XML from statically referenced assemblies may no longer be processed if the embedding assembly is not marked.
+
+- Constant propagation and branch elimination will will run per marked assembly
+
+  Unmarked assemblies will not be processed for constants and will not have unused branches removed. This will not inline constants across assemblies, and leaves room to optimize constant propagation to run at a more granular level inline with the marking logic.
+
+### Restrictions
+
+We will restrict the embedded XML so that it may only modify the containing assembly. This is to prevent cases where a (possibly lazily loaded) assembly modifies code in another assembly that we have already processed, breaking assumptions and creating logical inconsistencies.
+
+Embedded XML from assemblies which are not marked may not be processed, to avoid the need to load all referenced assemblies in case they have embedded XML.
+
+To support on-demand processing one assembly at a time, we will temporarily remove support for constant propagation across assembly boundaries. Note that to remove unused branches based on constant callees, we need to know whether callees - possibly in other assemblies - are constant. This would require loading and processing direct assembly references for constants before doing branch elimination. In general case, this could require recursive processing of assembly references for new constants introduced by branch elimination. There are plans to move constant propagation to `MarkStep`, which will add back cross-assembly constant propagation. 
+
+### Exceptions to lazy loading
+
+When the linker default action is `copy`, we will preserve the .NET5 behavior that keeps all statically referenced assemblies (that don't override the default action). For consistency with this behavior, static references of dynamically loaded assemblies will be kept as well. This will be done by pre-loading the reference closure of assemblies if the linker's default action is `copy`.
```

Comment:
> When the linker default action is `copy`

What does it mean? linker rarely runs in a single action mode

Reply:
I meant `-c`/`-u`. I'm struggling a bit with how this should work. Maybe it would be ok to remove support for `-c copy`/`-u copy` instead, though we probably need to figure out https://github.com/mono/linker/issues/1269 before making that decision (there was a suggestion to use `copy` as the default action).

Reply:
I think we will always need to load all referenced assemblies, to support opting in to trimming via attributes. Even if we don't do it by passing `-c copy -u copy`, we want everything without an attribute to be kept - so we need to look at references. Maybe we can do it without reading the entire assembly, so that we still get some perf savings from lazy loading.